### PR TITLE
Bug fix/failed to `HackAndSlash3` when `AvatarState` already max level and max exp

### PR DESCRIPTION
**Pull request by boscohyun**
- Bug fix/failed to `HackAndSlash3` when `AvatarState` already max level and max exp
- Improve equipment validation. (Prevent equipping more equipments than the slot that player has.)

### CODE REVIEW

File: Lib9c/Action/HackAndSlash4.cs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+using Nekoyume.TableData;
+using Serilog;
+
+namespace Nekoyume.Action
+{
+    [Serializable]
+    [ActionType("hack_and_slash4")]
+    public class HackAndSlash4 : GameAction
+    {
+        public List<int> costumes;
```
Comment (ipdae) [inline Lib9c/Action/HackAndSlash4.cs:21]: i think we can use `Guid` instead of `int`

File: Lib9c/Action/HackAndSlash4.cs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+                    $"Aborted due to insufficient action point: " +
+                    $"{avatarState.actionPoint} < {stageRow.CostAP}"
+                );
+            }
+
+            avatarState.actionPoint -= stageRow.CostAP;
+
+            avatarState.EquipCostumes(new HashSet<int>(costumes));
+
+            avatarState.EquipEquipments(equipments);
```
Comment (boscohyun) [inline Lib9c/Action/HackAndSlash4.cs:158]:
```suggestion
            var itemsToEquip = new List<Guid>(costumes).AddRange(equipments);
            avatarState.EquipItems(itemsToEquip);
```

https://github.com/planetarium/lib9c/blob/6eb98d42ccd4cc7bff86688af6bbb74e01c84d19/Lib9c/Model/State/AvatarState.cs#L522

**Review by ipdae — COMMENTED** [review]
test is broken

**Review by ipdae — APPROVED** [review]

**Comment by boscohyun** [comment]
Unit testing is required for all exception cases of `HackAndSlash4`.

**Review by boscohyun — COMMENTED** [review]
Unit testing is required for all exception cases of `HackAndSlash4`.

**Comment by hskim881028** [comment]
To replace with ValidateEquipmentsV2(),  MimisbrunnrBattle action also needs to make version 2.

**Review by ipdae — COMMENTED** [review]
i think copy logging code from `RankingBattle2`

**Comment by ipdae** [comment]
test is broken

**Review by Unengine — APPROVED** [review]

File: Lib9c/Action/RankingBattle3.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    public class RankingBattle3 : GameAction
+    {
+        public const int StageId = 999999;
+        public static readonly BigInteger EntranceFee = 100;
+
+        public Address AvatarAddress;
+        public Address EnemyAddress;
+        public Address WeeklyArenaAddress;
+        public List<Guid> equipments;
+        public List<Guid> equipmentIds;
```
Comment (boscohyun) [inline Lib9c/Action/RankingBattle3.cs:29]:
```suggestion
        public List<Guid> costumeIds;
        public List<Guid> equipmentIds;
```

File: Lib9c/Model/Character/Player.cs [hunk 4]
```diff
@@ -130,17 +130,17 @@ EquipmentItemSetEffectSheet equipmentItemSetEffectSheet
...
+            CharacterLevelSheet characterLevelSheet,
             EquipmentItemSetEffectSheet equipmentItemSetEffectSheet
         ) : base(
             null,
             characterSheet,
             GameConfig.DefaultAvatarCharacterId,
             level)
         {
-            Exp.Current = 0;
+            Exp.Current = characterLevelSheet[level].Exp;
```
Comment (boscohyun) [inline Lib9c/Model/Character/Player.cs:143]: These changes will cause a state mismatch. However, the `Player(int level, ...)` constructor only inherits from the `EnemyPlayer(int level, ...)` constructor, and this `EnemyPlayer(int level, ...)` constructor is not used. That's why I suggest deleting both constructors.

File: Lib9c/Model/State/AvatarState.cs [hunk 5]
```diff
@@ -423,6 +423,77 @@ public void ValidateEquipments(List<Guid> equipmentIds, long blockIndex)
             }
         }
 
+        public void ValidateEquipmentsV2(List<Guid> equipmentIds, long blockIndex)
```
Comment (boscohyun) [inline Lib9c/Model/State/AvatarState.cs:426]: We need unit tests for this method in `AvatarStateTest` class.

**Comment by hskim881028** [comment]
I think Log should be added to RankingBattle3 action.
this PR [link](https://github.com/planetarium/lib9c/pull/255#pullrequestreview-557722380)

File: .Lib9c.Tests/Action/RankingBattle3Test.cs [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+            using var ms = new MemoryStream();
+            formatter.Serialize(ms, action);
+            ms.Seek(0, SeekOrigin.Begin);
+
+            var deserialized = (RankingBattle3)formatter.Deserialize(ms);
+            Assert.Equal(action.PlainValue, deserialized.PlainValue);
+        }
+
+        [Theory]
+        [InlineData(ItemSubType.Weapon, GameConfig.MaxEquipmentSlotCount.Weapon)]
```
Comment (ipdae) [inline .Lib9c.Tests/Action/RankingBattle3Test.cs:397]: Add case for other equipments

File: Lib9c/Model/Character/Player.cs (hunk 4 above)
Comment (Unengine) [reply Lib9c/Model/Character/Player.cs:143]: These constructors are for tests!

**Review by hskim881028 — APPROVED** [review]

**Comment by ipdae** [comment]
test is broken

**Pull request merged**